File: abc286/e/main.py
N = int(input())
# N, K = map(int, input().split())
A = list(map(int, input().split()))

S = []
for i in range(N):
    S.append(input())

INF = float("inf")
d = [[INF]*N for i in range(N)]
val = [[0]*N for i in range(N)]
for i in range(N):
    d[i][i] = 0
    for j in range(N):
        val[i][j] = A[i]

for i in range(N):
    for j in range(N):
        if i == j:
            continue
        if S[i][j] == 'Y':
            d[i][j] = 1
            val[i][j] += A[j]

# 最短距離更新
for k in range(N): # 経由する点
    for i in range(N): # 始点
        for j in range(N): # 終点
            if d[i][k] == INF or d[k][j] == INF:
                continue
            if d[i][j] > d[i][k]+d[k][j]:
                d[i][j] = d[i][k]+d[k][j]
                val[i][j] = val[i][k]+val[k][j]-A[k]
            elif d[i][j] == d[i][k]+d[k][j]:
                val[i][j] = max(val[i][j], val[i][k]+val[k][j]-A[k])

Q = int(input())
for _ in range(Q):
    u, v = map(int, input().split())
    u, v = u-1, v-1
    if d[u][v] == INF:
        print('Impossible')
    else:
        print(d[u][v], val[u][v])

File: abc286/e/main.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <limits>
using namespace std;

const long long INF = numeric_limits<long long>::max();

int main() {
    int N;
    cin >> N;

    vector<long long> A(N);
    for (int i = 0; i < N; ++i) {
        cin >> A[i];
    }

    vector<string> S(N);
    for (int i = 0; i < N; ++i) {
        cin >> S[i];
    }

    vector<vector<long long>> d(N, vector<long long>(N, INF));
    vector<vector<long long>> val(N, vector<long long>(N, 0));

    // 初期化
    for (int i = 0; i < N; ++i) {
        d[i][i] = 0;
        for (int j = 0; j < N; ++j) {
            val[i][j] = A[i];
            if (S[i][j] == 'Y') {
                d[i][j] = 1;
                val[i][j] += A[j];
            }
        }
    }

    // 最短距離更新 (ワーシャル–フロイド法)
    for (int k = 0; k < N; ++k) { // 経由する点
        for (int i = 0; i < N; ++i) { // 始点
            for (int j = 0; j < N; ++j) { // 終点
                if (d[i][k] == INF || d[k][j] == INF) continue; // 無限大の場合はスキップ
                if (d[i][j] > d[i][k] + d[k][j]) {
                    d[i][j] = d[i][k] + d[k][j];
                    val[i][j] = val[i][k] + val[k][j] - A[k];
                } else if (d[i][j] == d[i][k] + d[k][j]) {
                    val[i][j] = max(val[i][j], val[i][k] + val[k][j] - A[k]);
                }
            }
        }
    }

    int Q;
    cin >> Q;
    while (Q--) {
        int u, v;
        cin >> u >> v;
        --u, --v; // 0-indexedに変換
        if (d[u][v] == INF) {
            cout << "Impossible" << endl;
        } else {
            cout << d[u][v] << " " << val[u][v] << endl;
        }
    }

    return 0;
}
